aQute.libg/src/aQute/libg/uri/URIUtil.java

```diff
@@ -41,4 +44,84 @@ public static URI resolve(URI baseURI, String reference) throws URISyntaxExcepti
 		return baseURI.resolve(refURI);
 	}
 
+	private static final URI EMPTYURI = URI.create("");
+
+	/**
+	 * Attempts to fetch a path on the file system for the given URI string.
+	 * Tries a few more tricks than the standard method of
+	 * <tt>Paths.get(new URI(uriString))</tt> - it can handle plain paths, and
+	 * also nested URI pseudo-schemes like <tt>reference:</tt> and
+	 * <tt>jar:</tt>. Examples:
+	 * <ul>
+	 * <li>/path/to/file => /path/to/file</li>
+	 * <li>reference:file:/path/to/file => /path/to/file</li>
+	 * <li>jar:file:/path/to/file.jar!/some/contained/element =>
+	 * /path/to/file.jar</li>
+	 * <li>http://server/path/to/file => <tt>null</tt></li>
+	 * </ul>
+	 *
+	 * @param uriString The string containing the URI for which we are
+	 *            attempting to construct a path.
+	 * @return The method's best guess as to which file on the local filesystem
+	 *         this URI refers, or <tt>null</tt> if it's invalid or not a local
+	 *         filesystem URI.
+	 */
+	public static Path fromURI(String uriString) {
+		if (uriString == null || uriString.isEmpty()) {
+			return null;
+		}
+
+		try {
+			URI uri = resolve(EMPTYURI, uriString);
+			while (uriString != null && !uriString.isEmpty()) {
+
+				final String scheme = uri.getScheme();
+				if (scheme == null) {
+					try {
+						return Paths.get(uriString);
+					} catch (InvalidPathException e) {
+						e.printStackTrace();
+						return null;
+					}
+				}
+
+				switch (uri.getScheme()
+					.toLowerCase()) {
+					case "file" :
+						return Paths.get(uri);
+					case "jar" :
+					case "bundle" :
+					case "zip" :
+						uriString = uri.getRawSchemeSpecificPart();
```

Comment:
Not sure why you are using `raw` methods in this method. I did not use them in `Resource.fromURL`.

Reply:
I vaguely thought if there was a path that had a : in it that had been URI-encoded, you wouldn't want that to be decoded as it might then be interpreted as a scheme delimiter when it shouldn't. But that's a pretty esoteric case. None of my test cases covered that case - as : is not a legal path character on Windows it wasn't easy for me to test. Maybe you could try some tests for files with ':' in the path on Mac and see whether it is appropriate to use `raw`.

Reply:
In the meantime, I've change it to not use the `raw`, for the sake of getting this merged.

Reply:
`:` is not generally use in *ix file names either since `:` is used as the path separator.

Reply:
Not generally used, yes; but it is not illegal - you can use it in a path if it's escaped. On Windows you simply can't use it at all.